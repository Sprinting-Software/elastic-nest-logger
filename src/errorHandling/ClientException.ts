import { AppException } from "./AppException"
import { HttpStatus } from "@nestjs/common"

/**
 * An exception using HTTP status code 500 InternalServerError.
 */
export class ClientException extends AppException {
  constructor(
    public readonly errorName: string,
    public readonly description?: string,
    public readonly contextData?: Record<string, any>,
    public readonly innerError?: Error
  ) {
    super(HttpStatus.BAD_REQUEST, errorName, description, contextData, innerError)
    Object.setPrototypeOf(this, ClientException.prototype)
  }
}
