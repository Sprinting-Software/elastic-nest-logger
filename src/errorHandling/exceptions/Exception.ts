import util from "util"
import { HttpStatus } from "@nestjs/common"
import { LibraryVersioning } from "../../libVersioning/LibraryVersioning"

export interface AppExceptionResponse {
  statusCode: number
  errorName: string
  errorTraceId: string
  message?: string | object
  contextData?: Record<string, any>
  innerError?: Error
}
export interface AppExceptionResponseV2 {
  httpStatus: number
  errorName: string
  errorTraceId: string
  message?: string | object
  contextData?: Record<string, any>
  innerError?: Error
}

const INSPECT_DEPTH = 7
const INSPECT_SHOW_HIDDEN = false
const MSG_LENGTH = 8445 // Max message length for UDP

export class Exception extends Error {
  public readonly errorTraceId: string
  constructor(
    public readonly httpStatus: number = HttpStatus.BAD_REQUEST,
    public errorName: string,
    public description?: string,
    public contextData: Record<string, any> = {},
    public innerError?: Error
  ) {
    super(errorName ?? HttpStatus[httpStatus])
    this.errorName = errorName ?? HttpStatus[httpStatus]
    this.errorTraceId = Exception.generateErrorTraceId()
    this.refreshMessageField()
    Object.setPrototypeOf(this, Exception.prototype)
  }

  public getStatus(): number {
    return this.httpStatus
  }
  override toString(): string {
    try {
      let msg = `${this.constructor.name} ${this.concatAllRelevantInfo()}`
      if (this.stack) msg += `\n ${this.generateStacktrace()}`
      if (this.innerError) {
        msg += "\nINNER_ERROR:\n"
        if (this.innerError instanceof Exception) {
          msg += this.innerError.toString()
        } else {
          const e0 = util.inspect(this.innerError)
          msg += e0
        }
      }
      if (msg.length > MSG_LENGTH) {
        return `${msg.substring(0, MSG_LENGTH)}...TRUNCATED`
      }
      return msg
    } catch (e) {
      return "ERROR_IN_TO_STRING"
    }
  }
  public generateStacktrace(): string {
    try {
      return this.stack?.split("\n").slice(1).join("\n")
    } catch (e) {
      return "STACKTRACE_GENERATION_FAILED"
    }
  }

  /**
   * Used to generate a string that can be used as a message field in ELK and in console output
   * @private
   */
  private concatAllRelevantInfo() {
    const { errorName, httpStatus, description, contextData } = this
    let msg = `${errorName} (HTTP_STATUS ${httpStatus}) (${this.errorTraceId}) `
    if (description) msg += `ERROR_DESCRIPTION - ${description}`
    if (contextData) msg += ` - CONTEXT_DATA: ${util.inspect(contextData, INSPECT_SHOW_HIDDEN, INSPECT_DEPTH)}`
    return msg
  }

  /**
   * We need to refresh the message field because we have to support the fluent API
   * where not all fields are necessarily passed in the constructor.
   * @private
   */
  private refreshMessageField() {
    this.message = this.concatAllRelevantInfo()
  }

  /**
   * Use this method to add an inner error to the current AppException object.
   * An inner error is an error that caused the current error to occur.
   *
   * @param {Error} innerError - The inner error to add to the AppException object.
   * @returns {Exception} - The modified AppException object with the new inner error.
   */
  setInnerError(innerError: Error): Exception {
    this.innerError = innerError
    this.refreshMessageField()
    return this
  }

  /*
   * Adds additional context data to the response object. Existing context data will be merged with the new data.
   * @param {Record<string, any>} contextData - The context data to be included with the error.
   * @returns {AppException} - The modified AppException object with the enriched context data
   */
  setContextData(contextData: Record<string, any>) {
    this.contextData = { ...this.contextData, ...contextData }
    this.refreshMessageField()
    return this
  }

  /**
   * Returns the response object that will be sent to the client.
   * Please do not change the method name as it matches with the NestJS built-in error interface.
   */
  getResponse(): AppExceptionResponse | AppExceptionResponseV2 {
    if (LibraryVersioning.v2IsActive()) {
      return {
        httpStatus: this.httpStatus,
        errorName: this.errorName,
        message: this.description,
        contextData: this.contextData,
        errorTraceId: this.errorTraceId,
      }
    } else {
      return {
        statusCode: this.httpStatus,
        errorName: this.errorName,
        message: this.description,
        contextData: this.contextData,
        errorTraceId: this.errorTraceId,
      }
    }
  }

  /**
   * Generates a random string that can be used as an error trace id.
   * Should only contain upper case letters that cannot be confused with numbers.
   * @private
   */
  private static generateErrorTraceId() {
    const result = new Array(6)
    function getRandom(cs: string = CHARS) {
      return cs.charAt(Math.floor(Math.random() * cs.length))
    }
    for (let i = 0; i < 6; i++) {
      result[i] = getRandom(CHARS)
    }
    return `ERR${result.join("")}`
  }
}
const CHARS = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
