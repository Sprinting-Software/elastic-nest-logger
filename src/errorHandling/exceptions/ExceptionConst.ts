let truncationLimit = 8445 // Max message length for UDP
export const ExceptionConst = {
  getTruncationLimitForExceptions: function () {
    return truncationLimit
  },
  overrideTruncationLimitForExceptions: function (limit: number) {
    truncationLimit = limit
  },
}
