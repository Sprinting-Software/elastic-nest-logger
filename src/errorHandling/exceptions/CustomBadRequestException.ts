import { BadRequestException } from "@nestjs/common"
import { Exception, AppExceptionResponse, AppExceptionResponseV2 } from "./Exception"
import { LibraryVersioning } from "../../libVersioning/LibraryVersioning"

export class CustomBadRequestException extends Exception {
  public readonly errors: Record<string, string[]> = {}

  constructor(private readonly badRequestException: BadRequestException) {
    super(400, badRequestException.name, badRequestException.message)

    Object.setPrototypeOf(this, CustomBadRequestException.prototype)
    this.errors = badRequestException.getResponse()["message"]
  }

  override toString(): string {
    const { errorName, httpStatus } = this
    let message = `${errorName} (HTTP_STATUS ${httpStatus})`
    message += ` ERROR_NAME - ${this.errorName}`
    message += ` ERROR_MESSAGE - ${this.description}`

    if (this.errors) message += ` VALIDATION_ERRORS - ${JSON.stringify(this.errors)}`

    return message
  }

  override getResponse() {
    if (LibraryVersioning.v2IsActive()) {
      return {
        httpStatus: this.httpStatus,
        errorName: this.errorName,
        message: this.errors,
      } as AppExceptionResponseV2
    } else {
      return {
        statusCode: this.httpStatus,
        errorName: this.errorName,
        message: this.errors,
      } as AppExceptionResponse
    }
  }
}
