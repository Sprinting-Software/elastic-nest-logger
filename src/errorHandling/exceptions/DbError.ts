import { Exception } from "./Exception"

export class DbError extends Exception {
  public operationName: string
  public args: any[]

  constructor(operationName: string, args: any[] = []) {
    super(500, `Database operation ${operationName} failed`)
    this.operationName = operationName
    this.args = args
    Object.setPrototypeOf(this, DbError.prototype)
  }
}
