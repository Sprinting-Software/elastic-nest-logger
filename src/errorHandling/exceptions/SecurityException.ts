import { AppExceptionResponse, AppExceptionResponseV2, Exception } from "./Exception"
import { HttpStatus } from "@nestjs/common"
import { LibraryVersioning } from "../../libVersioning/LibraryVersioning"

/**
 * An exception using HTTP status code 403 FORBIDDEN.
 * Exceptions of this type are treated specially in that the context data, stacktrace and description
 * are never returned to the client, not even in non-production.
 * This is to ensure a minimum amount of information leakage.
 */
export class SecurityException extends Exception {
  constructor(
    public readonly description?: string,
    public readonly contextData: Record<string, any> = {},
    public readonly innerError?: Error
  ) {
    super(HttpStatus.FORBIDDEN, "SecurityException", description, contextData, innerError)
    Object.setPrototypeOf(this, SecurityException.prototype)
  }

  /**
   * Make sure details about security errors are never returned to the client.
   */
  override getResponse(): AppExceptionResponse | AppExceptionResponseV2 {
    if (LibraryVersioning.v2IsActive()) {
      return {
        httpStatus: this.httpStatus,
        errorName: this.errorName,
        errorTraceId: this.errorTraceId,
      }
    } else {
      return {
        statusCode: this.httpStatus,
        errorName: this.errorName,
        errorTraceId: this.errorTraceId,
      }
    }
  }
}
