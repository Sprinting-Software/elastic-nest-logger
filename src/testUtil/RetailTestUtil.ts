import { ExceptionUtil } from "../errorHandling/ExceptionUtil"
import { AssertionException } from "../errorHandling/exceptions/AssertionException"
import { IExceptionJson } from "../errorHandling/IExceptionJson"

export class RetailTestUtil {
  /**
   * Takes any function that is expected to throw an error, catches the error and returns it as a plain json for
   * build test assertions.
   * If no error is thrown, an AssertionException is returned as plain json so that you are sure to handle the situation
   * where error handling does not happen as expected.
   * @param fct
   */
  public static catchAndReturnAsJson(fct: () => void) {
    try {
      fct()
      return ExceptionUtil.toPlainJsonForSpec(
        new AssertionException(
          "An exception should have been thrown at this point. There must be a coding error somewhere. "
        )
      )
    } catch (e) {
      return ExceptionUtil.toPlainJsonForSpec(e)
    }
  }

  /**
   * Same as catchAndReturnAsJson, only for async functions.
   * @param fct
   */
  public static async catchAndReturnAsJsonAsync(fct: () => Promise<void>): Promise<IExceptionJson> {
    try {
      await fct()
      return ExceptionUtil.toPlainJsonForSpec(
        new AssertionException(
          "An exception should have been thrown at this point. There must be a coding error somewhere. "
        )
      )
    } catch (e) {
      return ExceptionUtil.toPlainJsonForSpec(e)
    }
  }
}

/**
 * A short alias for ExceptionUtil.catchAndReturnAsJson
 */
export function CatchSync(fct: () => void) {
  return RetailTestUtil.catchAndReturnAsJson(fct)
}

/**
 * A short alias for ExceptionUtil.catchAndReturnAsJsonAsync
 */
export async function CatchAsync(fct: () => Promise<void>) {
  return await RetailTestUtil.catchAndReturnAsJsonAsync(fct)
}
