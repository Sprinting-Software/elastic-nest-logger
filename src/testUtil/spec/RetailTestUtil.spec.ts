import { CatchSync, CatchAsync } from "../RetailTestUtil"
import { ClientException } from "../../errorHandling/exceptions/ClientException"

describe("JwtHelper", () => {
  it("basic snapshot-testing", () => {
    const user = {
      createdAt: new Date(),
      createdOn: `str${new Date()}`,
      name: "Bond... James Bond",
    }

    expect(user).toMatchSnapshot({
      createdAt: expect.any(Date),
      createdOn: expect.any(String),
      name: "Bond... James Bond",
    })
  })

  it("exceptions and snapshot-testing sync", () => {
    const err = new ClientException("SomeErrorName", "some description", { a: 1, b: "xxx" }, new Error("InnerError"))

    function util() {
      throw err
    }

    expect(CatchSync(() => util())).toMatchSnapshot()
  })

  it("exceptions and snapshot-testing async", async () => {
    const err = new ClientException("SomeErrorName", "some description", { a: 1, b: "xxx" }, new Error("InnerError"))
    function utilAsync(): Promise<void> {
      return new Promise(function (resolve, reject) {
        reject(err)
      })
    }
    //await utilAsync()
    expect(await CatchAsync(async () => await utilAsync())).toMatchSnapshot()
  })
})
