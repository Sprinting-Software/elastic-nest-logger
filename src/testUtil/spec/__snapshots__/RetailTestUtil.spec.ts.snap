// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JwtHelper basic snapshot-testing 1`] = `
{
  "createdAt": Any<Date>,
  "createdOn": Any<String>,
  "name": "Bond... James Bond",
}
`;

exports[`JwtHelper exceptions and snapshot-testing async 1`] = `
{
  "contextData": {
    "a": 1,
    "b": "xxx",
  },
  "description": "some description",
  "errorName": "SomeErrorName",
  "errorTraceId": "REDACTED",
  "httpStatus": 400,
  "innerError": {
    "description": "InnerError",
    "errorName": "Error",
  },
}
`;

exports[`JwtHelper exceptions and snapshot-testing sync 1`] = `
{
  "contextData": {
    "a": 1,
    "b": "xxx",
  },
  "description": "some description",
  "errorName": "SomeErrorName",
  "errorTraceId": "REDACTED",
  "httpStatus": 400,
  "innerError": {
    "description": "InnerError",
    "errorName": "Error",
  },
}
`;
